### ISSUE TITLE
Provide a way to specify rastered content scale for transformed content

### ISSUE THREAD

**Issue by chrishtr** [w3c/csswg-drafts#236]
[css-transforms]

Currently, developers have no control over the scale at which to raster content with transforms. If the transform is not compositor-animated, there is only one logical answer (layout content scale plus transform). However, for performance reasons, compositor animations often work by scaling a pre-rastered bitmap in the GPU. This begs the question of the scale at which to raster that bitmap.

Each browser has its own heuristics for choosing the scale of these bitmaps, and when to re-raster them
when the scale changes. This is a problem when the heuristics fail, because the developer has no control
over details of performance and quality that are important for a high-quality rendered output.

Blink has recently done work to improve this situation. Some documents on what we're doing and background:

What we're doing: https://docs.google.com/a/chromium.org/document/d/1f8WS99F9GORWP_m74l_JfsTHgCrHkbEorHYu72D4Xag/edit?usp=drive_web&usp=docs_home&ths=true

Investigation of situation before the above and alternatives considered:
https://docs.google.com/document/d/1CsDfsMxZaM094VhTDrHXF86_rmFnpjM8xL3iEh7HWgg/edit#heading=h.omyv4h6h0sym

One thing that the above plans do not  cover is control over current-frame raster scale, if the desired such scale differs from the "ideal" scale (meaning layout scale + transform). There are cases where this is desired, for example rastering at a destination scale then animating to it, to maximize quality and performance. (Note that this is how Blink actually implements declarative CSS animation rastering today)
See also this blink-dev thread for examples and related discussion:
https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/Ufwrhx_iZ7U

**Comment by smfr** [comment]
I agree this would be a useful property. WebKit tries to do it automatically, and reverted the change because it had bad perf impacts. Putting this under author control would be good.

**Comment by tobireif** [comment]
Will this allow the web developer to express eg the following?

"This SVG element will change in scale by factor 3.4, browser please prepare for swift animation."
(The animation might happen later. And it could happen over a duration, going through the full range of sizes between scale 1 and scale 3.4 .)

It's just an example - it should also be usable with HTML not just SVG, and it should support all other relevant properties and values not just transform:scale.

**Comment by tobireif** [comment]
(I guess that would allow the browser to pre-rasterize all the required sizes -> not blurry and hopefully super-fast.)

**Comment by chrishtr** [comment]
> This really is necessary. Chrome's initial implementation of `will-change` for `transform` was so bad when it came to scale transforms that devs were dropping use of `will-change` completely.

If you can raise examples, that would be great. The "initial implementation" is definitely not what Chrome does right now.

I also posted detailed documents about what we've done since then to attempt to optimize, and reasons why we didn't pursue a raster scale spec change to date. (TL;DR: it seems hard to say what it means without over-fitting to implementation details and technologies)

**Comment by AmeliaBR** [comment]
> The "initial implementation" is definitely not what Chrome does right now.

Yes, things have improved.  But my comment was specifically in reference to the late 2016 / Chrome 53 behavior, which I assume is what is described in the links above from June 2016.  Discussion from that time: [Greensock blog](https://greensock.com/will-change), [my twitter thread](https://twitter.com/AmeliasBrain/status/781521012152266752) including the "will-change with range syntax" musings I referenced above. The example I link to in that thread no longer has the problem (JS scaling animation never re-rasterizing while `will-change: transform` is active). That is, Chrome now re-rasterizes when it scales up too much from the original.

Maybe that's enough & this issue is out of date. But I assume there are still going to be some JS-driven animations where it will be useful to give the rendering engine a heads up about what `will-change` is going to mean.

**Comment by chrishtr** [comment]
> The example I link to in that thread no longer has the problem (JS scaling animation never re-rasterizing while `will-change: transform` is active). That is, Chrome now re-rasterizes when it scales up too much from the original.

Right, thanks for mentioning that. We added a [tweak](https://chromium-review.googlesource.com/c/chromium/src/+/546580/) to not rasterize below "device" scale (meaning the "natural" scale of the composited layer). This was in response to an example that is a lot like the one in your Twitter thread. That made the Greensock folks happier, and may have resolved the most important use cases.

> Maybe that's enough & this issue is out of date. But I assume there are still going to be some JS-driven animations where it will be useful to give the rendering engine a heads up about what `will-change` is going to mean.

I agree with this. Right now it seems we've optimized the heuristics as well as we might be able to, and I'm sure some use cases are still not so great.

Maybe there could be a way of specifying in CSS that a property will be animated within the range of a declared keyframe sequence?

**Comment by flackr** [comment]
Would this allow developers to render downsampled content? A common technique to render a fast blur is to blur a downsampled version of the content and scale it up[1] but our tweak prevents this kind of optimization.

[1] "it is fairly common to sacrifice some quality and, since blurring is a low-pass filter (preserves low frequency signals while attenuating high frequency ones), down-scale to a smaller resolution buffer, for example to ½ by ½, perform the blur effect and then up-scale back to the original frame buffer": https://software.intel.com/en-us/blogs/2014/07/15/an-investigation-of-fast-real-time-gpu-based-image-blur-algorithms

**Comment by progers** [comment]
Does `<canvas>` provide a primitive for this? The canvas backing size (defined by the width and height attributes) is effectively the raster scale that will be used by `drawImage`. This technique has been used for pixel-perfect canvas on high-dpi displays, where the device pixel ratio is used as the raster scale. If the canvas is transformed (scaled) further, the `image-rendering` property controls the resampling algorithm used between the raster scale and the screen.

**Comment by tabatkins** [comment]
It doesn't, no. There's no equivalent to specify the backing-store size of an arbitrary element's painting. (And I don't think we'd want the direct equivalent either, as it would require authors to guess at the size of the element beforehand.)

I think something along the lines of an "ideal scale" that hints at the texture size the author wants is probably the direction we want here.

**Comment by Awendel** [comment]
If this issue is still relevant, 
I have suggested a new CSS property that would also cover the problem described here.

https://github.com/w3c/csswg-drafts/issues/7848

It is a lot simpler though, since it's just a multiplier to the existing layer resolution pipeline.
Because of its simplicity there would be fewer edge cases and concerns around implementation.

But combined with the "raster scale lock" via will-change: transform, developers would be able to "plan ahead" the content scale
 before an animation.

**Comment by heapwolf** [comment]
I think this is still relevant. A common case is where a div gets `transform: scale(4)`, and if that div contains a child with `overflow: scroll`, the parent becomes extremely pixelated - it would be great to be able to tell the browser to repaint the parent. I'd be interested in implementing a solution for Webkit. I like the `pixel-density` proposal, if there is no other existing solution (I'm assuming there isn't since this issue is still open).

**Comment by chanon** [comment]
I am developing an application that uses CSS animations heavily. 3D CSS animations in fact. The issue I have run into is that when underpowered GPUs are connected to high resolution screens, they can't keep up with all the animation updates and the framerate tanks. 

For example, an Intel UHD Graphics connected to a 4K display might handle general web browsing fine, but for my app rendering the whole animations in 4K is overkill and is beyond the GPU's capacity. However if I reduce the resolution to just HD (1920x1080), then it is a lot smoother and looks fine for its purpose.

A property like the proposed "pixel-density" above would be immensely useful and would even allow for adjusting the image-quality/performance balance as needed.
